Make plants use fertilizers described with json.

nbrichau opened the pull request:
This PR is the implementation of #750.
I replaced the fertilizer adapterizer with a fertilizer registry.
I also did some changes the api to use the new fertilizers.

I took inspiration of the way soils are registered to register the fertilizers. Hopefully I did it the right way.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizer.java:
```diff
@@ -1,18 +1,35 @@
+/**
+ * An interface for managing AgriCraft fertilizers.
+ */
 public interface IAgriFertilizer extends IAgriRegisterable<IAgriFertilizer> {
```
I am not 100% on the modifications here.
Functionally, the IAgriFertilizer has not changed, only the internal AgriCraft implementation.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizer.java:
```diff
@@ -1,18 +1,35 @@
 
+    @Nonnull
+    @Override
+    String getId();
```
Redundant.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizer.java:
```diff
@@ -1,18 +1,35 @@
+    ITextComponent getName();
+
+    @Nonnull
+    Collection<Item> getVariants();
```
This is not needed.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizer.java:
```diff
@@ -28,6 +45,19 @@
+     *
+     * @return the value of the potency of the fertilizer
+     */
+    int getPotency();
```
This is relevant for AgriCraft's implementation only, not for the abstract represenation of fertilizers.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizerProvider.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.Optional;
+
+@FunctionalInterface
+public interface IAgriFertilizerProvider {
```
Why do you phase out the Fertilizerr Adapterizer, but then introduce a Provider, which is functionally the same thing?

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizerRegistry.java:
```diff
@@ -0,0 +1,54 @@
+    void registerFertilizerProvider(@Nonnull Item item, @Nonnull IAgriFertilizerProvider fertilizerProvider);
+
+    @Nonnull
+    IAgriFertilizerProvider getProvider(@Nonnull Item item);
```
Not quite sure about the necessity of this.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/CoreHandler.java:
```diff
@@ -214,4 +215,26 @@ private static void initMutations() {
+        // Announce Progress
+        AgriCore.getLogger("agricraft").info("Registering Fertilizers!");
+
+        // See if soils are valid...
```
I suppose this is a ctrl+c / ctrl+v artefact?

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/CoreHandler.java:
```diff
@@ -214,4 +215,26 @@ private static void initMutations() {
+                .map(JsonFertilizer::new)
+                .forEach(AgriApi.getFertilizerRegistry()::add);
+
+        // Display Soils
```
Ditto

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/AgriFertilizerRegistry.java:
```diff
@@ -0,0 +1,68 @@
+
+    private static final AgriFertilizerRegistry INSTANCE = new AgriFertilizerRegistry();
+    private static final IAgriFertilizerProvider EMPTY = (item) -> Optional.empty();
+    private final Map<Item, IAgriFertilizerProvider> providers;
```
Instead of a map, we could keep the Adapterizer here, this does not restrict to Item -> fertilizer bijections, allowing for NBT-sensitive stuff as well.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/AgriFertilizerRegistry.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public boolean remove(@Nullable IAgriFertilizer element) {
+        // do not allow removal of the default no fertilizer implementation
+        return NO_FERTILIZER != element && super.remove(element);
```
Good.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -0,0 +1,97 @@
+        if((entity instanceof PlayerEntity) && !(((PlayerEntity) entity).isCreative())) {
+            stack.shrink(1);
+        }
+        // to spawn the bone meal particles
```
Shouldn't this use the json's particles now?

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -0,0 +1,97 @@
+        }
+        // to spawn the bone meal particles
+        world.playEvent(2005, pos, 0);
+        return ActionResultType.SUCCESS;
```
I believe the negative effect is missing?

InfinityRaider reviewed (changes requested):
Looks good overall, the only thing I am not sure about is the phasing out of the Adapterizer, in favor of a Provider, which are, functionally almost identical, except that the latter is more restrictive.

nbrichau commented:
I think all the changes needed are done. Some notable things since your review are :
- I replaced the fertilizer registry with the adapterizer.
- I removed the NoFertilizer (I don't see why it's usefull now that I use an adapterizer).
- I removed `IAgriPlant#isFertilizable`. I hesitated to deprecated the method, but because it isn't used anywhere in the code I removed it.
- `IAgriFertilizable#acceptsFertilizer` is now unused in the code. Every call are replaced by `IAgriFertilizer#canFertilize`.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -77,22 +77,25 @@ public boolean canTriggerWeeds() {
                         IAgriGrowthStage current = crop.getGrowthStage();
                         IAgriGrowthStage previous = current.getPreviousStage(crop, random);
-                        if(current.equals(previous)) {
+                        if (current.equals(previous)) {
```
Unsure whether or not to move the negative effect outside the loop 
For the positive effect it makes sense, as not every growth tick is guaranteed to pass 
The negative effect is however applied every pass, so I feel the negative effect is disproportionately more powerful than the positive one.

nbrichau replied:
I think that more a fertilizer is powerful, more its negative effect should be powerful, and easier it gets to kill the plant.
The negative effect have half a chance to be applied (L 83 : `&& random.nextBoolean()`). So I don't think it's that disproportionate, but I could change the probability to 1/3 or 1/4.

InfinityRaider replied:
Ah I see, I must have read over that random check.

nbrichau commented:
I figured that this was a bad decision, and some checks are made in `IAgriFertilizable#acceptsFertilizer` (in agricraft's implementation) that wheren't done after I changed that.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/fertilizer/IAgriFertilizer.java:
```diff
@@ -11,7 +12,10 @@
+/**
+ * An interface for managing AgriCraft fertilizers.
+ */
+public interface IAgriFertilizer extends IAgriRegisterable<IAgriFertilizer>, IAgriAdapter<IAgriFertilizer> {
```
Is the IAfriRegistrable still needed?

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/api/v1/plant/IAgriPlant.java:
```diff
@@ -72,14 +72,6 @@ default ItemStack toItemStack(int amount) {
-     * @return If the plant can be fertilized.
-     */
-    boolean isFertilizable(IAgriGrowthStage growthStage, IAgriFertilizer fertilizer);
-
```
This should be kept and called.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/content/core/TileEntityCropBase.java:
```diff
@@ -374,7 +374,7 @@ public boolean acceptsFertilizer(@Nonnull IAgriFertilizer fertilizer) {
         } else if(this.hasPlant()) {
-            return (!this.isFullyGrown()) && this.getPlant().isFertilizable(this.getGrowthStage(), fertilizer);
+            return (!this.isFullyGrown()) && fertilizer.canFertilize(this);
         } else {
```
Needs to call the plant method as well.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -0,0 +1,164 @@
+    @Override
+    public ActionResultType applyFertilizer(World world, BlockPos pos, IAgriFertilizable target, ItemStack stack, Random random, @Nullable LivingEntity entity) {
+        for (int i = 0; i < this.potency; i++) {
+            if (target instanceof IAgriCrop && ((IAgriCrop) target).hasPlant()) {
```
Tthe hasPlant() check will prevent fertilizer usage on cross crops to force mutations.

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -0,0 +1,164 @@
+        if (!(target instanceof IAgriCrop)) {
+            return false;
+        }
+        return ((IAgriCrop) target).hasPlant() && this.fertilizerEffect.canFertilize(((IAgriCrop) target).getPlant().getId());
```
What about fertilizing cross crops to force mutations?

InfinityRaider commented on src/main/java/com/infinityraider/agricraft/impl/v1/fertilizer/JsonFertilizer.java:
```diff
@@ -0,0 +1,164 @@
+
+    private void spawnParticles(IAgriFertilizable target, Random rand) {
+        if (!(target instanceof IAgriCrop)) {
+            return;
```
I don't like this cast at each time but there isn't really a way around it with how we have the json set up....
